api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,324 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+    vscode.LanguageModelChatMessage.User(codeWithLineNumbers),
+  ];
+
+  // make sure the model is ready
+  if (model) {
+
+    // send the messages array to the model and get the response
+    let chatResponse = await model.sendRequest(messages, {}, new vscode.CancellationTokenSource().token);
+
+  }
+});
+```
+
+Then we'll send the messages to the model and get the response. Chat responses come in as fragments. These fragments usually contain single words, but sometimes they contain just punctuation. In order to display annotations as the response streams in, we want to wait until we have a complete annotation before we display it. Because of the way we have instructed our model to return its response, we know that when we see a closing `}` we have a complete annotation. We can then parse the annotation and display it in the editor.
+
+Add the following function above the `getCodeWithLineNumbers` method in the `extension.ts` file.
+
+```ts
+async function parseChatResponse(chatResponse: vscode.LanguageModelChatResponse, textEditor: vscode.TextEditor) {
+ let accumulatedResponse = "";
+
+ for await (const fragment of chatResponse.text) {
+  accumulatedResponse += fragment;
+
+  // if the fragment is a }, we can try to parse the whole line
+  if (fragment.includes("}")) {
+   try {
+    const annotation = JSON.parse(accumulatedResponse);
+    applyDecoration(textEditor, annotation.line, annotation.suggestion);
+    // reset the accumulator for the next line
+    accumulatedResponse = "";
+   }
+   catch (e) {
+    // do nothing
+   }
+  }
+ }
+}
+```
+
+Now we can call this method from the `code-tutor.annotate` command. Modify the implementation of the command so that it looks like this:
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+
+  // init the chat message
+  const messages = [
+    vscode.LanguageModelChatMessage.User(ANNOTATION_PROMPT),
+    vscode.LanguageModelChatMessage.User(codeWithLineNumbers),
+  ];
+
+  // make sure the model is ready
+  if (model) {
+
+    // send the messages array to the model and get the response
+    let chatResponse = await model.sendRequest(messages, {}, new vscode.CancellationTokenSource().token);
+
+    // parse the response and display the annotations
+    await parseChatResponse(chatResponse, textEditor);
+
+  }
+});
+```
+
+We need one last method to actually display the annotations. VS Code calls these "decorations". Add the following method above the `parseChatResponse` method in the `extension.ts` file.
+
+```ts
+function applyDecoration(editor: vscode.TextEditor, line: number, suggestion: string) {
```

Comment:
I suggest to call the variable `textEditor` to align with other parts of the code.